rfc/rfc-80/rfc-80.md

```diff
@@ -0,0 +1,161 @@
+
+Constraints: The column family list must contain the primary key, and columns contained in different column families cannot overlap except for the primary key. The preCombie field does not need to be specified. If not specified, the primary key will be taken by default.
+
+After the table is created, the column family attributes will be persisted to hoodie's metadata for subsequent use.
+
+### Adding and deleting column families in existing table
+Use the SQL alter command to modify the column family attributes and persist it:    
+* Execute ALTER TABLE table_name SET TBLPROPERTIES ('hoodie.columnFamily.k'='a,b,c;a'); to add a new column family.  
+* Execute ALTER TABLE table_name UNSET TBLPROPERTIES('hoodie.columnFamily.k'); to delete the column family.
+
+Specific steps are as follows:
+1. Execute the ALTER command to modify the column family
+2. Verify whether the column family modified by alter is legal. Column family modification must meet the following conditions, otherwise the verification will not pass:
+    * The column family name of an existing column family cannot be modified.  
+    * Columns in other column families cannot be divided into new column families.  
+    * When creating a new column family, it must meet the format requirements from previous chapter.  
+3. Save the modified column family to the .hoodie directory.
+
+### Writing data
+The Hudi kernel divides the input data according to column families; the data belonging to a certain column family is sorted and directly written to the corresponding column family log file.
+
+![process-write](process-write.png)
+
+Specific steps:  
+1. The engine divides the written data into buckets according to hash and shuffles the data (the writing engine completes it by itself and is consistent with the current writing of the native bucket).  
+2. The Hudi kernel sorts the data to be written to each bucket by primary key.  
+3. After sorting, split the data into column families.  
+4. Write the segmented data into the log file of the corresponding column family.  
+
+#### Common API interface
+After the table columns are clustered, the writing process includes the process of sorting and splitting the data compared to the original bucket bucketing. A new append interface needs to be introduced to support column families.  
+Introduce ColumnFamilyAppendHandle extend AppendHandle to implement column family writing.
+
+![append-handle](append-handle.png)
+
+### Reading data
+#### ColumnFamilyReader and RowReader
+![row-reader](row-reader.png)
+
+Hudi internal row reader reading steps:  
+1. Hudi organizes files by column families to be read.
+2. Introduce familyReader to merge and read each column family's own baseFile and logfile to achieve column family-level data reading.  
+    * Since log files are written after being sorted by primary key, familyReader merges its own baseFile and logFile by primary key using sortMerge.
```

Comment:
So the compaction would also utilize the reader right?

Reply:
Yes, exactly.